chess-AI/mcts.py

```diff
@@ -0,0 +1,176 @@
+from math import sqrt
+
+import numpy as np
+
+from state_representation import get_cnn_input
+
+class Mcts:
+    """ Allows for a  Monte
+    Carlo Tree Search to be performed
+
+
+    Attributes:
+        states_visited: List storing the states visited
+        n_values: Dictionary storing the n values
+        q_values: Dictionary storing the q values
+        exploration: The exploration rate for the tree search
+        p_values: Dictionary storing the search probabilites  
+    """
+
+    def __init__(self, exploration):
+        self.states_visited = [] 
+        self.n_values = {} 
+        self.q_values = {}
+        self.exploration = exploration 
+        self.p_values = {} 
+
+    def find_best_move(self, legal_moves, fen_string):
+
+        """
+        Find and return the best move given the current
+        state and legal moves
+        """
+
+        best_u = -float("inf")
+        best_move = -1
+
+        for move in legal_moves:
+
+            uci_move = move.uci()
+
+            #Iterate over every legal move at current board state.
+            #Initialize q_value if state has not been visited, 
+            #otherwise get q_value.
+
+            if fen_string in self.q_values:
+                if uci_move in self.q_values[fen_string]:
+                    q_value = self.q_values[fen_string][uci_move]
+                else:
+                    self.q_values[fen_string][uci_move] = 0
+                    q_value = 0
+            else:
+                self.q_values[fen_string] = {}
+                self.q_values[fen_string][uci_move] = 0
+                q_value = 0
```

Comment:
This and the code below code can maybe be abstracted into a single function? Depends on your initialization of n values. We can talk about it more, but this might allow less duplicated code

```python
def set_current_qn_value(self, values, fen_string, uci_move, initial_value):
    if fen_string in values:
        if uci_move in self.values[fen_string]:
            return self.values[fen_string][uci_move]
        else:
            self.values[fen_string][uci_move] = initial_value
            return initial_value
    else:
        self.values[fen_string] = {}
        self.values[fen_string][uci_move] = initial_value
        return initial_value
```

then you can change this and the lower block to
```python
q_value = self.set_current_qn_value(self.q_values, fen_string, uci_move, initial_value=0)
n_value = self.set_current_qn_value(self.n_values, fen_string, uci_move, initial_value=1)
```